### ISSUE TITLE
Inflatet Type I Error rate for feature importance

### ISSUE THREAD

**Issue by rajoma-02** [citoverse/cito#70]
The Type I Error rate of the feature importance seems to be inflated:
I simulated 100 datasets with two predictors (X1 and X2). X1 has an effect of 1 on the dependent variable Y, X2 has no effect. Each dataset contains 100 observations. The p-values are estimated through 50 bootstrapped simulations. I got 46% significant p-values for the feature with no effect (feature X2). The Type-I-Error rate is correct for ACE, but inflated for the feature importance

Code:

```
#function to create one dataset
create_data_mat <- function(n_obs=300, means=c(-1,1), sds=c(1,2), effs=c(1,0)){ # KISS
  if(any(c(length(sds), length(means))!=length(effs))) stop("equal number of means, standard deviations and effects is needed")
  X<-matrix(nrow = n_obs, ncol =length(means))
  for(i in 1:length(means)){
    X[, i]<-rnorm(n_obs, means[i], sds[i])
  }
  eps<-rnorm(n_obs)
  
  Y<-X%*%effs+eps
  
  data<-data.frame(Y,X)
  return(data)
}

#function to extract the p-values
get_p_vals_dnn<-function(n_sim=100, n_obs=300, means=c(0,0), sds=c(1,1), effs=c(1,0), n_boots=50){
  library(cito)
  pvals_mat_f_imp<-matrix( ncol=length(means), nrow =0)
  pvals_mat_ACE<-matrix( ncol=length(means), nrow =0)
  
  
  for(i in 1:n_sim){
    print(paste(round(i/n_sim, digits = 2)*100, "%"))
    data<- create_data_mat(n_obs=n_obs, means=means, sds=sds, effs=effs)
    fit<-dnn(Y~., data =data, verbose = F, plot = F, bootstrap = n_boots, epochs = 10)
    summ<-summary(fit)
    pvals_mat_ACE<-rbind(pvals_mat_ACE,summ$ACE_coefmat[,4])
    pvals_mat_f_imp<-rbind(pvals_mat_f_imp,summ$importance_coefmat[,4])
  }
  res<-c(apply(pvals_mat_ACE, 2 ,function(x){length(which(x<0.05))/n_sim}),
         apply(pvals_mat_f_imp, 2 ,function(x){length(which(x<0.05))/n_sim}))
  names(res)<-c(paste0("pval_ACE_X", 1:length(means)),paste0("pval_f_imp_X", 1:length(means)))
  return(res)
}
a<-get_p_vals_dnn()
print(a)
```

**Comment by florianhartig** [comment]
As discussed, this is very likely not a technical error, but just reflects that, due to overfitting, a variable will have in-sample importance even if there is no true effect.

A possible remedy (to be checked) would be to calculate the importance under CV.

This is analogue to adjusted / unadjusted R2 in ANOVA. 

We need to decide what to do with this. Minimal would probably be a remark in the help?

**Comment by MaximilianPi** [comment]
> A possible remedy (to be checked) would be to calculate the importance under CV.
> 
> This is analogue to adjusted / unadjusted R2 in ANOVA.
> 
> We need to decide what to do with this. Minimal would probably be a remark in the help?

I agree, but doing a CV after bootstrapping is computationally impractical since we have to re-fit all the models... 
Why not use the out-of-bag samples , we have them and they are free. A quick preliminary analysis:

Type I error rates for the importance:

On the training data (X1 has an effect, X2 has no effect):
X1 → X2 →  
 1.00 0.86 

On out-of-bag:
X1 → X2 →  
 1.00 0.00 


Not really at 5%, a thorough test needs to be done, but I think it is a promising start.

Distribution of the p-values for oob of X2:
<img width="741" alt="image" src="https://github.com/citoverse/cito/assets/29951520/12994e03-f4fd-4bc8-800f-15414f46e594">

**Comment by florianhartig** [comment]
OK, CV is not impossible, but would increase runtime by folds, right? 

I agree it's a problem and would be better to avoid that but need to check if this works with the out-of-bag samples. I'm surprised that type I is 0, would have expected >0.05 because of lacking independence.

**Comment by florianhartig** [comment]
OK, seems out-of-bag works well and I take the comment back about a lack of independence, I guess we'll go ahead with that!
